File: src/main/java/uk/gov/london/ops/repository/UserRepository.java
/**
 * Copyright (c) Greater London Authority, 2016.
 *
 * This source code is licensed under the Open Government Licence 3.0.
 *
 * http://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/
 */
package uk.gov.london.ops.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import uk.gov.london.ops.domain.user.User;

import java.util.Set;

public interface UserRepository extends JpaRepository<User, String> {



    @Query(value = "select * from users us " +
            "left outer join user_roles ur on us.username = ur.username and ur.primary_org_for_user = true " +
            "and primary_org_for_user = null", nativeQuery = true)
    Set<User> findAllUsersWithoutPrimaryRole();
}

File: gla-ops-application/src/main/java/uk/gov/london/ops/user/implementation/UserRepository.kt
package uk.gov.london.ops.user.implementation

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import uk.gov.london.ops.user.domain.UserEntity
import java.util.*

interface UserRepository : JpaRepository<UserEntity?, String?> {
    fun findByUserId(userId: Int?): UserEntity?

    @Query(value = "select distinct u.* from users u "
            + " join user_roles ur on u.username = ur.username "
            + " where ur.name in :roles",  nativeQuery = true)
    fun findByUserRoles(@Param("roles") roles: List<String>) : List<UserEntity>


    @Query(value = "select u.username from users u where u.enabled = true and u.password_expiry < ?1",  nativeQuery = true)
    fun findAllUsernamesExpiredBefore(cutoffDate: Date): Set<String>

    @Query(value = "select u.* from users u inner join user_roles ur on u.username = ur.username " +
            "where ur.organisation_id = ?1 and ur.name in ?2", nativeQuery = true)
    fun findByOrganisationIdAndUserRoles(organisationId: Int?, vararg roles: String?): Set<UserEntity?>?
}
